File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/InitialGpsEventFactory.java
package com.mapbox.services.android.navigation.v5.internal.navigation;

import com.mapbox.core.utils.TextUtils;

class InitialGpsEventFactory {

  private static final String EMPTY_STRING = "";
  private String sessionId = EMPTY_STRING;
  private ElapsedTime time;
  private InitialGpsEventHandler handler;
  private boolean hasSent;

  InitialGpsEventFactory() {
    this(new ElapsedTime(), new InitialGpsEventHandler());
  }

  InitialGpsEventFactory(ElapsedTime time, InitialGpsEventHandler handler) {
    this.time = time;
    this.handler = handler;
  }

  void navigationStarted(String sessionId) {
    this.sessionId = sessionId;
    time.start();
  }

  void gpsReceived() {
    if (time.getStart() == null) {
      return;
    }
    time.end();
    send(time);
  }

  void reset() {
    time = new ElapsedTime();
    hasSent = false;
  }

  private void send(ElapsedTime time) {
    if (!hasSent && !TextUtils.isEmpty(sessionId)) {
      double elapsedTime = time.getElapsedTime();
      handler.send(elapsedTime, sessionId);
      hasSent = true;
    }
  }
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/InitialGpsEventFactory.kt
package com.mapbox.services.android.navigation.v5.internal.navigation

import com.mapbox.navigation.utils.time.ElapsedTime
import com.mapbox.services.android.navigation.v5.navigation.metrics.MetricsReporter

internal class InitialGpsEventFactory @JvmOverloads constructor(
    metricsReporter: MetricsReporter,
    private var time: ElapsedTime = ElapsedTime(),
    private val handler: InitialGpsEventHandler = InitialGpsEventHandler(metricsReporter)
) {

    private var sessionId = ""
    private var hasSent = false

    fun navigationStarted(sessionId: String) {
        this.sessionId = sessionId
        time.start()
    }

    fun gpsReceived(metadata: NavigationPerformanceMetadata) {
        if (time.start == null) {
            return
        }
        time.end()
        send(time, metadata)
    }

    fun reset() {
        time = ElapsedTime()
        hasSent = false
    }

    private fun send(time: ElapsedTime, metadata: NavigationPerformanceMetadata) {
        if (!hasSent && sessionId.isNotEmpty()) {
            val elapsedTime = time.elapsedTime
            handler.send(elapsedTime, sessionId, metadata)
            hasSent = true
        }
    }
}
